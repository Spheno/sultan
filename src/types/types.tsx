export interface ICard {
  id: number;
  title: string;
  url: string;
  sizeType: string;
  size: string;
  barcode: number;
  manufacturer: string;
  brand: string;
  subtitle: string;
  price: number;
  description: string;
  useFor: string;
  quantity?: number;
  inBasket?: boolean;
  typeCare: ITypeCare;
}

export interface ITypeCare {
  body?: boolean;
  hand?: boolean;
  feet?: boolean;
  face?: boolean;
  hair?: boolean;
  sun?: boolean;
  shaving?: boolean;
  gift?: boolean;
  hygiene?: boolean;
  oral?: boolean;
  paper?: boolean;
}