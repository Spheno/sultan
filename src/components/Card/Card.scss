@import "../../scss/constants";

.card {
  width: 326px;
  @extend %flexCenter;
  flex-direction: column;
  padding: 29px 23px;
  box-shadow: 0 15px 70px rgba(0, 0, 0, 0.1);
  border-radius: 10px;
}

.card__img-container {
  width: 220px;
  height: 220px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
}

.card__img {
  max-width: 220px;
  max-height: 220px;
}

.card__type-weight {
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
  margin: 4px 0 10px;

}

.card__weight {
  background-position: center;
  background-repeat: no-repeat;
  width: 20px;
  height: 16px;
  margin-right: 5px;
}

.card__weight_weight {
  background-image: url('../../images/box.svg');
}

.card__weight_volume {
  background-image: url('../../images/bottle.svg');
}

.card__weight-text {
  font-size: 12px;
  font-weight: 400;
  line-height: 1.3;
  color: $color-2;
}

.card__title {
  font-weight: 700;
  font-size: 16px;
  line-height: 1.3;
  color: $color-1;
  @extend %overflowTitle;
  cursor: pointer;

  span {
    font-weight: 500;
  }

}

.card__specification-container {
  @extend %flexStart;
  row-gap: 5px;
  margin: 20px 0 15px;
}

.card__specification {
  font-size: 14px;
  font-weight: 300;
  line-height: 1.5;
  color: $color-2;

  span {
    font-weight: 500;
    color: $color-1;
  }

}

.card__price-container {
  width: 100%;
  @extend %flexBetween;
}

.card__price {
  font-size: 16px;
  font-weight: 800;
  line-height: 1.3;
  color: $color-1;
}

.card__button {
  @extend %fontButton;
  @extend %button;
  color: $mainBackgroundColor;
  width: 153px;
  text-transform: uppercase;
  padding: 16px 25px;
  background-image: url('../../images/basket-white.svg');
  background-position: 100px 10px;
  background-size: 27px 27px;
  border-radius: 80px;
  font-size: 10px;
  font-weight: 700;
  line-height: 1.2;
  letter-spacing: 1.5px;
  text-align: start;
}

:disabled {
  background-color: $activeColor;
  background-image: none;
  text-align: center;
  cursor: default;
}

//*****************************************************************

@media (max-width: 780px) {

  .card {
    width: 290px;
    box-shadow: 0 9px 32px rgba(0, 0, 0, 0.1);
  }

}