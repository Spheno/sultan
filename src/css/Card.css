.card__price-container {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.card {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.card__specification-container {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  -webkit-box-pack: start;
      -ms-flex-pack: start;
          justify-content: flex-start;
  -webkit-box-align: start;
      -ms-flex-align: start;
          align-items: flex-start;
}

.card__button {
  font-size: 14;
  font-weight: 700;
  line-height: 1.5;
  color: #fff;
}

.card__button {
  background-color: #FFC85E;
  border: none;
  background-repeat: no-repeat;
  border-radius: 79px;
}

.card__title {
  overflow: hidden;
  -o-text-overflow: ellipsis;
     text-overflow: ellipsis;
  display: -moz-box;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  line-clamp: 2;
  box-orient: vertical;
  margin-bottom: 10px;
}

.card {
  width: 326px;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
  padding: 29px 23px;
  -webkit-box-shadow: 0 15px 70px rgba(0, 0, 0, 0.1);
          box-shadow: 0 15px 70px rgba(0, 0, 0, 0.1);
  border-radius: 10px;
}

.card__img-container {
  width: 220px;
  height: 220px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
      -ms-flex-pack: center;
          justify-content: center;
  -webkit-box-align: end;
      -ms-flex-align: end;
          align-items: flex-end;
}

.card__img {
  max-width: 220px;
  max-height: 220px;
}

.card__type-weight {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
      -ms-flex-pack: start;
          justify-content: flex-start;
  -webkit-box-align: start;
      -ms-flex-align: start;
          align-items: flex-start;
  margin: 4px 0 10px;
}

.card__weight {
  background-position: center;
  background-repeat: no-repeat;
  width: 20px;
  height: 16px;
  margin-right: 5px;
}

.card__weight_weight {
  background-image: url("../../images/box.svg");
}

.card__weight_volume {
  background-image: url("../../images/bottle.svg");
}

.card__weight-text {
  font-size: 12px;
  font-weight: 400;
  line-height: 1.3;
  color: #3F4E65;
}

.card__title {
  font-weight: 700;
  font-size: 16px;
  line-height: 1.3;
  color: #111;
  cursor: pointer;
}
.card__title span {
  font-weight: 500;
}

.card__specification-container {
  row-gap: 5px;
  margin: 20px 0 15px;
}

.card__specification {
  font-size: 14px;
  font-weight: 300;
  line-height: 1.5;
  color: #3F4E65;
}
.card__specification span {
  font-weight: 500;
  color: #111;
}

.card__price-container {
  width: 100%;
}

.card__price {
  font-size: 16px;
  font-weight: 800;
  line-height: 1.3;
  color: #111;
}

.card__button {
  color: #fff;
  width: 153px;
  text-transform: uppercase;
  padding: 16px 25px;
  background-image: url("../../images/basket-white.svg");
  background-position: 100px 10px;
  background-size: 27px 27px;
  border-radius: 80px;
  font-size: 10px;
  font-weight: 700;
  line-height: 1.2;
  letter-spacing: 1.5px;
  text-align: start;
}

:disabled {
  background-color: #1FD85D;
  background-image: none;
  text-align: center;
  cursor: default;
}

@media (max-width: 780px) {
  .card {
    width: 290px;
    -webkit-box-shadow: 0 9px 32px rgba(0, 0, 0, 0.1);
            box-shadow: 0 9px 32px rgba(0, 0, 0, 0.1);
  }
}