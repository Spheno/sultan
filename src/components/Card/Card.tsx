import './Card.scss';
import { useNavigate } from 'react-router-dom';
import { FC } from 'react';

import { ICard } from '../../types/types';

interface CardProps {
  product: ICard;
  inBasket: boolean;
  onCardClick: (card: ICard) => void;
  onButtonClick: (card: ICard) => void;
}

export const Card: FC<CardProps> = ({ product, onCardClick, onButtonClick, inBasket }) => {

  const navigate = useNavigate();

  const handleCardClick = () => {
    navigate(`/catalog/product/${product.id}`);
  };

  return (
    <div className="card">

      <div className="card__img-container">
        <img className="card__img" alt={product.title} src={product.url} />
      </div>

      <div className="card__info">

        <div className="card__type-weight">
          <div className={`card__weight ${product.sizeType === 'weight' ? `card__weight_weight` : `card__weight_volume`}`}></div>
          <p className="card__weight-text">{product.size}</p>
        </div>

        <h3 className="card__title" onClick={() => { onCardClick({ ...product, inBasket: inBasket }); handleCardClick() }}>{product.title} <span>{product.subtitle}</span></h3>

        <div className="card__specification-container">
          <p className="card__specification">Штрихкод: <span>{product.barcode}</span></p>
          <p className="card__specification">Производитель: <span >{product.manufacturer}</span></p>
          <p className="card__specification">Бренд: <span >{product.brand}</span></p>
        </div>

        <div className="card__price-container">
          <p className="card__price">{product.price} ₸</p>
          <button
            className="card__button" aria-label="Добавить в корзину" type="button"
            disabled={inBasket ? true : false} onClick={() => onButtonClick({ ...product, quantity: 1 })}>
            {inBasket ? 'В корзине' : 'В корзину'}
          </button>
        </div>
      </div>
    </div>
  )
}