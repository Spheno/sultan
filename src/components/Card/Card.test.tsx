import { Card } from './Card';
import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ICard } from '../../types/types';
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockedUsedNavigate,
}));

const product: ICard = {
  id: 1001,
  title: "BioMio BIO-SOAP",
  url: "https://www.podrygka.ru//upload/iblock/966/966a27334c6c8e6907571c0b9a0963d1.jpg",
  sizeType: "weight",
  size: "90г",
  barcode: 4604049097548,
  manufacturer: "BioMio",
  brand: "BioMio",
  subtitle: "Экологичное туалетное мыло. Литсея и бергамот",
  price: 50,
  description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam interdum ut justo, vestibulum sagittis iaculis iaculis. Quis mattis vulputate feugiat massa vestibulum duis. Faucibus consectetur aliquet sed pellentesque consequat consectetur congue mauris venenatis. Nunc elit, dignissim sed nulla ullamcorper enim, malesuada.",
  typeCare: [0, 1, 2, 3, 6],
  quantity: 3,
}

const onCardClick = jest.fn()
const onButtonClick = jest.fn()

describe('Cards', () => {

  it('button view if the product is in the cart', () => {
    render(<Card product={product} onCardClick={onCardClick} onButtonClick={onButtonClick} inBasket={true} />)
    const button = screen.getByRole('button')
    expect(button).toBeInTheDocument();
    fireEvent.click(button)
    expect(screen.getByRole('button')).toHaveTextContent('В корзине')
  })

  it('button view if the product is not in the cart', () => {
    render(<Card product={product} onCardClick={onCardClick} onButtonClick={onButtonClick} inBasket={false} />)
    const button = screen.getByRole('button')
    expect(button).toBeInTheDocument();
    fireEvent.click(button)
    expect(screen.getByRole('button')).toHaveTextContent('В корзину')
  })

  it('snapshot', () => {
    const view = render(
    <Card product={product} onCardClick={onCardClick} onButtonClick={onButtonClick} inBasket={false} />
    )

    expect(view).toMatchSnapshot()
  })

})