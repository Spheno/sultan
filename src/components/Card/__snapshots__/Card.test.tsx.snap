// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <div
          class="card__img-container"
        >
          <img
            alt="BioMio BIO-SOAP"
            class="card__img"
            src="https://www.podrygka.ru//upload/iblock/966/966a27334c6c8e6907571c0b9a0963d1.jpg"
          />
        </div>
        <div
          class="card__info"
        >
          <div
            class="card__type-weight"
          >
            <div
              class="card__weight card__weight_weight"
            />
            <p
              class="card__weight-text"
            >
              90г
            </p>
          </div>
          <h3
            class="card__title"
          >
            BioMio BIO-SOAP
             
            <span>
              Экологичное туалетное мыло. Литсея и бергамот
            </span>
          </h3>
          <div
            class="card__specification-container"
          >
            <p
              class="card__specification"
            >
              Штрихкод: 
              <span>
                4604049097548
              </span>
            </p>
            <p
              class="card__specification"
            >
              Производитель: 
              <span>
                BioMio
              </span>
            </p>
            <p
              class="card__specification"
            >
              Бренд: 
              <span>
                BioMio
              </span>
            </p>
          </div>
          <div
            class="card__price-container"
          >
            <p
              class="card__price"
            >
              50
               ₸
            </p>
            <button
              aria-label="Добавить в корзину"
              class="card__button"
              type="button"
            >
              В корзину
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <div
        class="card__img-container"
      >
        <img
          alt="BioMio BIO-SOAP"
          class="card__img"
          src="https://www.podrygka.ru//upload/iblock/966/966a27334c6c8e6907571c0b9a0963d1.jpg"
        />
      </div>
      <div
        class="card__info"
      >
        <div
          class="card__type-weight"
        >
          <div
            class="card__weight card__weight_weight"
          />
          <p
            class="card__weight-text"
          >
            90г
          </p>
        </div>
        <h3
          class="card__title"
        >
          BioMio BIO-SOAP
           
          <span>
            Экологичное туалетное мыло. Литсея и бергамот
          </span>
        </h3>
        <div
          class="card__specification-container"
        >
          <p
            class="card__specification"
          >
            Штрихкод: 
            <span>
              4604049097548
            </span>
          </p>
          <p
            class="card__specification"
          >
            Производитель: 
            <span>
              BioMio
            </span>
          </p>
          <p
            class="card__specification"
          >
            Бренд: 
            <span>
              BioMio
            </span>
          </p>
        </div>
        <div
          class="card__price-container"
        >
          <p
            class="card__price"
          >
            50
             ₸
          </p>
          <button
            aria-label="Добавить в корзину"
            class="card__button"
            type="button"
          >
            В корзину
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
